packages/flutter/lib/src/material/scaffold.dart

```diff
@@ -124,22 +200,12 @@ class _ScaffoldLayout extends MultiChildLayoutDelegate {
 
     if (hasChild(_ScaffoldSlot.floatingActionButton)) {
       final Size fabSize = layoutChild(_ScaffoldSlot.floatingActionButton, looseConstraints);
-      double fabX;
-      assert(textDirection != null);
-      switch (textDirection) {
-        case TextDirection.rtl:
-          fabX = _kFloatingActionButtonMargin + endPadding;
-          break;
-        case TextDirection.ltr:
-          fabX = size.width - fabSize.width - _kFloatingActionButtonMargin - endPadding;
-          break;
-      }
-      double fabY = contentBottom - fabSize.height - _kFloatingActionButtonMargin;
-      if (snackBarSize.height > 0.0)
-        fabY = math.min(fabY, contentBottom - snackBarSize.height - fabSize.height - _kFloatingActionButtonMargin);
-      if (bottomSheetSize.height > 0.0)
-        fabY = math.min(fabY, contentBottom - bottomSheetSize.height - fabSize.height / 2.0);
-      positionChild(_ScaffoldSlot.floatingActionButton, new Offset(fabX, fabY));
+      
+      // To account for the FAB position being changed, we'll lerp between its old and new position.
+      final Offset currentFabOffset = currentFabPosition.getOffset(size, fabSize, snackBarSize, bottomSheetSize, textDirection, contentBottom, endPadding);
+      final Offset previousFabOffset = previousFabPosition.getOffset(size, fabSize, snackBarSize, bottomSheetSize, textDirection, contentBottom, endPadding);
+      final Offset fabOffset = new MaterialPointArcTween(begin: previousFabOffset, end: currentFabOffset).lerp(fabMoveProgress);
```

Comment:
creating a tween each time seems unfortunate, we could just pass in the actual offset here if we moved the animation logic out to the caller

Reply:
I've refactored as much of the tween creation as possible.

I still need to create the exact offset tween in the scaffold layout delegate because the positions still need to respond to changes in, for example, the bottom sheet or snackbar during the animation.